orttraining/orttraining/core/graph/training_op_defs.cc

```diff
@@ -1387,6 +1387,37 @@ void RegisterTrainingOpSchemas() {
         }
       });
 
+  /**
+   * ClipGradNormInplace operator, taking multiple gradients as inputs (seq<tensor>).
+   * ClipGradNormInplace should be used in conjunction with optimizers that are expecting seq<tensor> gradients as input,
+   * since this op takes input a sequence of tensors and outputs a sequence of tensors there by avoiding the need
+   * for SequenceConstruct (and making any unnecessary copy).
+   * Please note that the gradient clipping happens inplace.
+   */
+  ONNX_CONTRIB_OPERATOR_SCHEMA(ClipGradNormInplace)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "gradients", "Sequence of gradients computed in this iteration.", "S_GRAD")
+      .Output(0, "clipped_gradients", "Gradients after being clipped as per given inputs and attributes.", "S_GRAD")
```

Comment:
Since the updates are in place, mark this output as optional?

Reply:
In this case, the output cannot be optional. The gradients that are updated in place, need to be fed into the folloing node (AdamWOptimizer). Without this context, the planner may place execution of the AdamWOptimizer ahead of ClipGradNorm which might lead to the wrong result.
One way to circumvent this is to set the execution priority for the clip grad norm as higher than AdamWOptimizer. Another is to feed the output into AdamWOptimizer.

Alternatively, we could make this as a non inplace op. The only problem with that is that the gradients are user owned. So, the user might not have visibility into the clipped gradients.

This is the reason we may not be able to make the output optional.